src/Interpreters/InterpreterSelectQueryAnalyzer.cpp

```diff
@@ -226,6 +233,11 @@ BlockIO InterpreterSelectQueryAnalyzer::execute()
     if (!select_query_options.ignore_quota && select_query_options.to_stage == QueryProcessingStage::Complete)
         result.pipeline.setQuota(context->getQuota());
 
+
+    /// Increase the SelectQueriesWithPrimaryKey profile.
```

Comment:
I see three problems with metric `SelectQueriesWithPrimaryKey` that will be hard to solve:

1. File InterpreterSelectQueryAnalyzer.cpp is the most central place of the Analyzer component of ClickHouse. Before this PR it was not cluttered with any ProfileEvent updates, now it is. ProfileEvent updates for similar counters (the ones in l. 10-17 of ProfileEvents.cpp) are implemented in Planner.cpp. Why is that not the case for SelectQueriesWithPrimaryKey?
2. Looking at the names and meaning (according to the comments) of the profile events in l. 10 - 17, they are all about counting queries of a specific type like SELECT and INSERT. Counting the number of queries that use the primary key breaks the scheme and introduces an inconsistency. The counter should be renamed to "SelectQueriesWithPrimaryKeyUsage" and moved before "AsyncInsertQuery". That is also not great but at least a little improvement.
3. The usage of the Context to store the metric result is unusual, error-prone, and further clutters the already too messy Context. We should keep it simple and not store ProfileEvents in the context. As far as I see, the context is required to get a per-query usage statistics (i.e. across multiple accessed parts) for the query. Could we not simply increase the metric by 1 in l. 1642 of ReadFromMergeTree.cpp and not update the context? The value will then be > 1 for a query but I think that is still okay?